rts/Sim/Weapons/Weapon.cpp

```diff
@@ -1275,13 +1276,241 @@ float3 CWeapon::GetUnitLeadTargetPos(const CUnit* unit) const
 	return aimPos;
 }
 
+// debug print includes
+//#include<iostream>
+//#include<string>
+
+float CWeapon::GetSafeInterceptTime(const CUnit* unit, float predictMult) const
+{
+	ZoneScoped;
+	float3 unitSpeed = unit->speed * predictMult;
+	float3 dist = unit->pos - weaponMuzzlePos;
+	float aa = unitSpeed.dot(unitSpeed) - (weaponDef->projectilespeed) * (weaponDef->projectilespeed);
+	float bb = 2 * (dist.dot(unitSpeed));
+	float cc = dist.dot(dist);
+	float temp1 = 4 * aa * cc;
+	float temp2 = (bb * bb);
+	float predictTime = 0.0;
+	// goal here is to return the smallest positive solution to a quadratic formula
+	// while also being numerically stable
+	// We also know c is strictly positive.
+
+	// case 1, aa <0, target speed is less than projectile speed
+	// guaranteed existence of a positive solution
+	// case 1a, aa is a large value, standard quadratic formula works fine
+	if (aa < -1) {
+		// standard quadratic formula
+		predictTime = (-bb - math::sqrt(temp2 - temp1)) / (2 * aa);
+	}
+	// case 1b, aa is a small value, "inverted" standard quadratic formula works better, and extra check needed
+	else if (aa <= 0) {
+		// check catastrophic case of aa=0 and bb=0 
+		// target speed equal to projectile speed, and target is moving tangentally
+		if ((std::abs(aa) < (std::pow(2, -16))) && (std::abs(bb) < (std::pow(2, -16)))) {
+			return -1.0;
+		}
+		// use Citardauq Formula if aa is small
+		predictTime = (2 * cc) / (-bb + math::sqrt(temp2 - temp1));
+	}
```

Comment:
What about the no-solution condition of `aa=0 and bb>0`? This makes the denominator `-bb + sqrt( bb*bb - 4*aa*cc ) = - bb + sqrt( bb^2 - 0 ) = - bb + bb = 0`, which means we divide by zero and get Infinity. It doesn't happen with `aa=0 and bb<0` as it reduces to `- (- bb) + sqrt ( (- bb)^2 ) = bb + sqrt ( bb^2 ) = bb + bb = 2 * bb`